File: DSA/DSA05026 XEM PHIM.cpp
#include <bits/stdc++.h>
using namespace std;

int check(vector<int> &A, int n, int k)
{
    vector<bool> temp(k+1, false);
    temp[0] = true;
    
    for (int i = 0; i < n; i++)
    {
        for (int j = k; j >= A[i]; j--)
        {
            if (temp[j-A[i]])
            {
                temp[j] = true;
            }
        }
    }
    
    for (int i = k - 1; i >= 0; i--)
    {
        if (temp[i])
        {
            return i;
        }
    }
}

void submit()
{
    int k, n;
    cin >> k >> n;
    vector<int> A(n);
    for (auto &&i : A)
    {
        cin >> i;
    }
    
    cout << check(A, n, k);
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    submit();

    return 0;
}

File: DSA_python/DSA05026 XEM PHIM.py
def submit():
    k, n = map(int, input().split())
    A = [int(input()) for _ in range(n)]

    def check():
        temp = [False]*(k+1)
        temp[0] = True

        for i in range(n):
            for j in range(k, A[i]-1, -1):
                if temp[j-A[i]] == True:
                    temp[j] = True
        for i in range(k-1, -1, -1):
            if temp[i] == True:
                return i
            
    print(check())


submit()
